### PR TITLE
Factor out Nullable support into a new type

### PR DESCRIPTION
**Pull request by cjp39**
Handling of nullable properties in DefaultPropertyFactory was over-eager code sharing. Splitting out a separate NullablePropertyFactory has simplified the code, and incidentally simplified the other property factories, which no longer need to rule out nullability before continuing. Analyser, Metadata and CodeGenerator no longer have custom logic to handle Nullable annotations, instead deferring to NullablePropertyFactory and its code generator.

Also added a reflection-based implementation of hashCode, equals and toString to the abstract type PropertyCodeGenerator for use in tests.

### CODE REVIEW

File: src/test/java/org/inferred/freebuilder/processor/NullablePropertyFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+@RunWith(MockitoJUnitRunner.class)
+public class NullablePropertyFactoryTest {
+
+  @Rule public final ModelRule model = new ModelRule();
+  @Mock(answer = RETURNS_SMART_NULLS) private Config config;
+  private final Property property = new Property.Builder().buildPartial();
+  private final NullablePropertyFactory factory = new NullablePropertyFactory();
+
+  @Before
+  public void setup() {
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullablePropertyFactoryTest.java]: Codacy Issue found: [You may have misspelled a JUnit framework method (setUp or tearDown)](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828624/issues/source?bid=3146243&fileBranchId=3187161#l51)

File: src/test/java/org/inferred/freebuilder/processor/NullablePropertyFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    when(config.getProperty()).thenReturn(property);
+  }
+
+  @Test
+  public void notNullable() {
+    ExecutableElement getterMethod = (ExecutableElement) model.newElementWithMarker(
+        "package com.example;",
+        "public class DataType {",
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullablePropertyFactoryTest.java]: Codacy Issue found: [The String literal "public class DataType {" appears 4 times in this file; the first occurrence is on line 59](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828624/issues/source?bid=3146243&fileBranchId=3187161#l59)

File: src/test/java/org/inferred/freebuilder/processor/NullablePropertyFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  @Before
+  public void setup() {
+    when(config.getProperty()).thenReturn(property);
+  }
+
+  @Test
+  public void notNullable() {
+    ExecutableElement getterMethod = (ExecutableElement) model.newElementWithMarker(
+        "package com.example;",
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullablePropertyFactoryTest.java]: Codacy Issue found: [The String literal "package com.example;" appears 4 times in this file; the first occurrence is on line 58](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828624/issues/source?bid=3146243&fileBranchId=3187161#l58)

File: src/test/java/org/inferred/freebuilder/processor/NullablePropertyFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        "public class DataType {",
+        "  ---> public abstract String getName();",
+        "  public static class Builder extends DataType_Builder {}",
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullablePropertyFactoryTest.java]: Codacy Issue found: [The String literal "  public static class Builder extends DataType_Builder {}" appears 4 times in this file; the first occurrence is on line 61](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828624/issues/source?bid=3146243&fileBranchId=3187161#l61)

File: src/main/java/org/inferred/freebuilder/processor/Analyser.java [hunk 5]
```diff
@@ -495,13 +483,16 @@ private PropertyCodeGenerator createCodeGenerator(
 
   private class ConfigImpl implements Config {
 
-    final Property property;
-    final Set<String> methodsInvokedInBuilderConstructor;
+    private final Property property;
+    private final ExecutableElement getterMethod;
+    private final Set<String> methodsInvokedInBuilderConstructor;
```
Comment (cjp39) [inline src/main/java/org/inferred/freebuilder/processor/Analyser.java]: Codacy Issue found: [Avoid excessively long variable names like methodsInvokedInBuilderConstructor](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828561/issues/source?bid=3146243&fileBranchId=3187161#l488)

File: src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+        "    @Override",
+        "    @Nullable",
+        "    public String getName() {",
+        "      return name;",
+        "    }",
+        "",
+        "    @Override",
+        "    @Nullable",
+        "    public Integer getAge() {",
+        "      return age;",
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java]: Codacy Issue found: [The String literal "      return age;" appears 4 times in this file; the first occurrence is on line 159](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828626/issues/source?bid=3146243&fileBranchId=3187161#l159)

File: src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+        "    @Override",
+        "    public int hashCode() {",
+        "      return Arrays.hashCode(new Object[] {name, age});",
+        "    }",
+        "",
+        "    @Override",
+        "    public String toString() {",
+        "      return \"Person{\"",
+        "          + COMMA_JOINER.join(",
+        "              (name != null ? \"name=\" + name : null), " +
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java]: Codacy Issue found: [The String literal "              (name != null ? \"name=\" + name : null), " appears 4 times in this file; the first occurrence is on line 186](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828626/issues/source?bid=3146243&fileBranchId=3187161#l186)

File: src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+        "    @Nullable private final Integer age;",
+        "",
+        "    private Value(Person_Builder builder) {",
+        "      this.name = builder.name;",
+        "      this.age = builder.age;",
+        "    }",
+        "",
+        "    @Override",
+        "    @Nullable",
+        "    public String getName() {",
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java]: Codacy Issue found: [The String literal "    public String getName() {" appears 4 times in this file; the first occurrence is on line 152](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828626/issues/source?bid=3146243&fileBranchId=3187161#l152)

File: src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+        "      return name;",
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java]: Codacy Issue found: [The String literal "      return name;" appears 4 times in this file; the first occurrence is on line 153](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828626/issues/source?bid=3146243&fileBranchId=3187161#l153)

File: src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+        "    public int hashCode() {",
+        "      return Arrays.hashCode(new Object[] {name, age});",
+        "    }",
+        "",
+        "    @Override",
+        "    public String toString() {",
+        "      return \"Person{\"",
+        "          + COMMA_JOINER.join(",
+        "              (name != null ? \"name=\" + name : null), " +
+            "(age != null ? \"age=\" + age : null))",
```
Comment (cjp39) [inline src/test/java/org/inferred/freebuilder/processor/NullableSourceTest.java]: Codacy Issue found: [The String literal "(age != null ? \"age=\" + age : null))" appears 4 times in this file; the first occurrence is on line 187](https://www.codacy.com/app/Chris-Purcell-39/FreeBuilder/file/2277828626/issues/source?bid=3146243&fileBranchId=3187161#l187)

File: src/main/java/org/inferred/freebuilder/processor/Analyser.java [hunk 11]
```diff
@@ -103,6 +103,7 @@
    * factory should always be last.
    */
   private static final List<PropertyCodeGenerator.Factory> PROPERTY_FACTORIES = ImmutableList.of(
+      new NullablePropertyFactory(), // Must be first
```
Comment (j-baker) [inline src/main/java/org/inferred/freebuilder/processor/Analyser.java]: why?

File: src/main/java/org/inferred/freebuilder/processor/Analyser.java [hunk 12]
```diff
@@ -111,7 +112,7 @@
       new SetMultimapPropertyFactory(),
       new OptionalPropertyFactory(),
       new BuildablePropertyFactory(),
-      new DefaultPropertyFactory());
+      new DefaultPropertyFactory()); // Must be last
```
Comment (j-baker) [inline src/main/java/org/inferred/freebuilder/processor/Analyser.java]: why?

File: src/main/java/org/inferred/freebuilder/processor/NullablePropertyFactory.java [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+import org.inferred.freebuilder.processor.Metadata.Property;
+import org.inferred.freebuilder.processor.PropertyCodeGenerator.Config;
+import org.inferred.freebuilder.processor.util.SourceBuilder;
+
+import java.util.Set;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.TypeElement;
+
+/** Default {@link PropertyCodeGenerator.Factory}, providing reference semantics for any type. */
```
Comment (j-baker) [inline src/main/java/org/inferred/freebuilder/processor/NullablePropertyFactory.java]: ??

**Comment by j-baker** [comment]
looks ok apart from the three comments

**Comment by cjp39** [comment]
Fixed, thanks! (Had to rebase after merging the other 3 PRs; the last commit is the fixes you asked for.)

**Pull request merged**